File: omarchy-plugin/lib/pomo_link/rest.py
"""Phone REST via urllib. Auth: X-Pomo-Token."""

from __future__ import annotations

import json
import urllib.error
import urllib.request

from .constants import HTTP_FLUSH_TIMEOUT_S, HTTP_TIMEOUT_S


class RestClient:
    def __init__(self):
        self.host = ""
        self.port = 0
        self.token = ""

    def configure(self, host, port, token):
        self.host = host or ""
        self.port = int(port or 0)
        self.token = token or ""

    def url(self, path, host=None, port=None):
        h = host if host is not None else self.host
        p = int(port if port is not None else self.port)
        if ":" in h and not h.startswith("["):
            h = "[%s]" % h
        return "http://%s:%d%s" % (h, p, path)

    def request(self, method, path, body=None, timeout=None, host=None, port=None, token=None):
        """Return (code, text). code 0 means transport failure."""
        timeout = HTTP_TIMEOUT_S if timeout is None else timeout
        tok = self.token if token is None else token
        data = None
        headers = {"X-Pomo-Token": tok or ""}
        if body is not None:
            if isinstance(body, (dict, list)):
                raw = json.dumps(body, separators=(",", ":")).encode("utf-8")
            elif isinstance(body, str):
                raw = body.encode("utf-8")
            else:
                raw = body
            data = raw
            headers["Content-Type"] = "application/json"
        if method == "POST" and data is None:
            data = b""
        req = urllib.request.Request(
            self.url(path, host=host, port=port),
            data=data,
            headers=headers,
            method=method,
        )
        try:
            with urllib.request.urlopen(req, timeout=timeout) as resp:
                text = resp.read().decode("utf-8", "replace")
                return int(resp.status), text
        except urllib.error.HTTPError as exc:
            try:
                text = exc.read().decode("utf-8", "replace")
            except Exception:
                text = ""
            return int(exc.code), text
        except Exception:
            return 0, ""

    def get_status(self, host=None, port=None, token=None, timeout=None):
        return self.request(
            "GET",
            "/api/status",
            timeout=timeout,
            host=host,
            port=port,
            token=token,
        )

    def get_config(self):
        return self.request("GET", "/api/config")

    def post(self, path, body=None, timeout=None):
        if timeout is None and path in ("/api/sessions/import", "/api/timer/adopt"):
            timeout = HTTP_FLUSH_TIMEOUT_S
        return self.request("POST", path, body=body, timeout=timeout)

File: extension/src/link/rest.ts
import { HTTP_FLUSH_TIMEOUT_MS, HTTP_TIMEOUT_MS } from "./constants";
import { phoneOrigin, type Pairing } from "./pairing";

export interface RestResult {
  status: number;
  body: string;
}

export interface RestPort {
  configure(pairing: Pairing): void;
  getStatus(): Promise<RestResult>;
  getConfig(): Promise<RestResult>;
  getHistory(): Promise<RestResult>;
  post(path: string, body?: unknown): Promise<RestResult>;
}

function parseJson(body: string): unknown {
  if (body.trim().length === 0) return null;
  try {
    return JSON.parse(body) as unknown;
  } catch {
    return null;
  }
}

export function jsonOf(result: RestResult): unknown {
  return parseJson(result.body);
}

export class PhoneRest implements RestPort {
  private baseUrl = "";
  private token = "";

  constructor(private readonly fetchImpl: typeof fetch = fetch) {}

  configure(pairing: Pairing): void {
    this.baseUrl = phoneOrigin(pairing);
    this.token = pairing.token;
  }

  getStatus(): Promise<RestResult> {
    return this.request("GET", "/api/status");
  }

  getConfig(): Promise<RestResult> {
    return this.request("GET", "/api/config");
  }

  getHistory(): Promise<RestResult> {
    return this.request("GET", "/api/history", undefined, HTTP_FLUSH_TIMEOUT_MS);
  }

  post(path: string, body?: unknown): Promise<RestResult> {
    const timeout = path === "/api/sessions/import" || path === "/api/timer/adopt" ? HTTP_FLUSH_TIMEOUT_MS : HTTP_TIMEOUT_MS;
    return this.request("POST", path, body, timeout);
  }

  private async request(method: string, path: string, body?: unknown, timeoutMs = HTTP_TIMEOUT_MS): Promise<RestResult> {
    if (this.baseUrl.length === 0) return { status: 0, body: "" };
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeoutMs);
    const headers: Record<string, string> = { "X-Pomo-Token": this.token };
    let payload: string | undefined;
    if (body !== undefined) {
      payload = typeof body === "string" ? body : JSON.stringify(body);
      headers["Content-Type"] = "application/json";
    }
    try {
      const response = await this.fetchImpl(`${this.baseUrl}${path}`, {
        method,
        headers,
        body: method === "POST" ? (payload ?? "") : undefined,
        signal: controller.signal,
      });
      return { status: response.status, body: await response.text() };
    } catch (error) {
      console.warn("pomo link fetch failed", method, path, error);
      return { status: 0, body: "" };
    } finally {
      clearTimeout(timer);
    }
  }
}

export interface SocketHandlers {
  onOpen: () => void;
  onMessage: (text: string) => void;
  onClose: () => void;
}

export interface SocketHandle {
  send(text: string): void;
  close(): void;
}

export type SocketFactory = (url: string, handlers: SocketHandlers) => SocketHandle;

export const browserSocket: SocketFactory = (url, handlers) => {
  const ws = new WebSocket(url);
  let closed = false;
  const notifyClose = (): void => {
    if (closed) return;
    closed = true;
    handlers.onClose();
  };
  ws.addEventListener("open", () => handlers.onOpen());
  ws.addEventListener("message", (event) => handlers.onMessage(String(event.data)));
  ws.addEventListener("close", notifyClose);
  ws.addEventListener("error", notifyClose);
  return {
    send: (text) => {
      if (ws.readyState === WebSocket.OPEN) ws.send(text);
    },
    close: () => {
      try {
        ws.close();
      } catch {
        // Already closed.
      }
    },
  };
};
